amlb/container.py

```diff
@@ -0,0 +1,170 @@
+"""
+**container** module is build on top of **benchmark** module to provide logic to create and run docker images
+that are preconfigured with a given automl framework, and that can be used to run a benchmark anywhere.
+The docker image embeds a version of the automlbenchmark app so that tasks are later run in local mode inside docker,
+providing the same parameters and features allowing to import config and export results through mounted folders.
+"""
+from abc import abstractmethod
+import logging
+import os
+import re
+
+from .benchmark import Benchmark, SetupMode
+from .errors import InvalidStateError
+from .job import Job
+from .resources import config as rconfig, get as rget
+from .utils import dir_of, run_cmd
+
+
+log = logging.getLogger(__name__)
+
+
+class ContainerBenchmark(Benchmark):
+    """ContainerBenchmark
+    an extension of Benchmark to run benchmarks inside a container.
+    """
+
+    @staticmethod
+    def image_name(framework_def, branch=None):
+        di = framework_def.image
+        if branch is None:
+            branch = rget().project_info.branch
+        return "{author}/{image}:{tag}".format(
+            author=di.author,
+            image=di.image if di.image else framework_def.name.lower(),
+            tag=re.sub(r"([^\w.-])", '.',
+                       '-'.join([di.tag if di.tag else framework_def.version.lower(), branch]))
+        )
+
+    @abstractmethod
+    def __init__(self, framework_name, benchmark_name, constraint_name):
+        """
+
+        :param framework_name:
+        :param benchmark_name:
+        :param constraint_name:
+        """
+        super().__init__(framework_name, benchmark_name, constraint_name)
+
+    def _validate(self):
+        if self.parallel_jobs == 0 or self.parallel_jobs > rconfig().max_parallel_jobs:
+            log.warning("Forcing parallelization to its upper limit: %s.", rconfig().max_parallel_jobs)
+            self.parallel_jobs = rconfig().max_parallel_jobs
+
+    def setup(self, mode, upload=False):
+        if mode == SetupMode.skip:
+            return
+
+        if mode == SetupMode.auto and self._image_exists():
+            return
+
+        self._generate_script(self.custom_commands)
+        self._build_image(cache=(mode != SetupMode.force))
+        if upload:
+            self._upload_image()
+
+    def cleanup(self):
+        # TODO: remove generated script? anything else?
+        pass
+
+    def run(self, task_name=None, fold=None):
+        self._get_task_defs(task_name)  # validates tasks
+        if self.parallel_jobs > 1 or not self.minimize_instances:
```

Comment:
let's try not to have abstractions relying on methods/properties defined only in implementations.
You can initialize those properties to defaults
```
self.minimize_instances = rconfig().container.minimize_instances
self.force_branch = rconfig().container.force_branch
```
in this abstract class, and let the implementations override them in their constructor.

Also, to avoid duplications, I'd rather have a `container` namespace in `config.yaml` before `docker` and `singularity` ones:
```
container: &container
  force_branch: true  # set to true if image can only be built from a clean branch, with same tag as defined in `project_repository`.
  minimize_instances: true
  run_extra_options: ''
  image:      # set this value through -container.image=my-image to run benchmark with a specific image
  image_defaults:
    author: automlbenchmark
    image:    # set by container impl based on framework name, lowercase
    tag:      # set by container impl based on framework version

docker:
  <<: *container
  run_extra_options: '--shm-size=1024M'

singularity:
  <<: *container
  library: 'automlbenchmark/default' 
```